File: src/algorithms/trees/properties/count-complete-tree-nodes/__tests__/count-complete-tree-nodes_test.go
package main

import "testing"

func makeCCTNNode(value int, left *TreeNode, right *TreeNode) *TreeNode {
	return &TreeNode{value: value, left: left, right: right}
}

func cctnLeaf(value int) *TreeNode {
	return &TreeNode{value: value}
}

func TestCountCompleteTreeNodes7Node(t *testing.T) {
	root := makeCCTNNode(4,
		makeCCTNNode(2, cctnLeaf(1), cctnLeaf(3)),
		makeCCTNNode(6, cctnLeaf(5), cctnLeaf(7)))
	if countCompleteTreeNodes(root) != 7 {
		t.Error("7-node perfect tree should have 7 nodes")
	}
}

func TestCountCompleteTreeNodesNull(t *testing.T) {
	if countCompleteTreeNodes(nil) != 0 {
		t.Error("null root should return 0")
	}
}

func TestCountCompleteTreeNodesSingleNode(t *testing.T) {
	if countCompleteTreeNodes(cctnLeaf(1)) != 1 {
		t.Error("single node should return 1")
	}
}

func TestCountCompleteTreeNodes3Node(t *testing.T) {
	root := makeCCTNNode(1, cctnLeaf(2), cctnLeaf(3))
	if countCompleteTreeNodes(root) != 3 {
		t.Error("3-node perfect tree should have 3 nodes")
	}
}

File: src/algorithms/trees/properties/count-complete-tree-nodes/CountCompleteTreeNodes_test.cpp
// g++ -o cctn_test CountCompleteTreeNodes_test.cpp && ./cctn_test
#include "sources/CountCompleteTreeNodes.cpp"
#include <cassert>
#include <iostream>

TreeNode* makeCCTNNode(int value, TreeNode* left = nullptr, TreeNode* right = nullptr) {
    TreeNode* node = new TreeNode(value);
    node->left = left;
    node->right = right;
    return node;
}

int main() {
    // test: 7-node perfect tree
    TreeNode* tree1 = makeCCTNNode(4,
        makeCCTNNode(2, makeCCTNNode(1), makeCCTNNode(3)),
        makeCCTNNode(6, makeCCTNNode(5), makeCCTNNode(7)));
    assert(countCompleteTreeNodes(tree1) == 7);

    // test: null root returns 0
    assert(countCompleteTreeNodes(nullptr) == 0);

    // test: single node
    assert(countCompleteTreeNodes(makeCCTNNode(1)) == 1);

    // test: 3-node perfect tree
    assert(countCompleteTreeNodes(makeCCTNNode(1, makeCCTNNode(2), makeCCTNNode(3))) == 3);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
